File: cpp/3063.linked-list-frequency.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* solve(ListNode* head) {
	    unordered_map<int, int> freq;
	    auto current = head;
	    while(current){
		    freq[current->val]++;
		    current = current->next;
	    }
	    ListNode * prev = NULL, *newHead = NULL;
	    current = NULL;
	    for(auto [val, count]:freq){
		    current = new ListNode(count);
		    if(!newHead)newHead = current;
		    if(prev)prev->next = current;
		    prev = current;
	    }

	    return newHead;
    }
    ListNode* frequenciesOfElements(ListNode* head) {
	    return solve(head);
    }
};

File: python/3063.linked-list-frequency.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def frequenciesOfElements(self, head: Optional[ListNode]) -> Optional[ListNode]:
        freq = {}
        current = head
        while current:
            freq[current.val] = freq.get(current.val, 0)+1
            current = current.next

        prev = None
        current = None
        prevHead = None
        for key, val in freq.items():
            current = ListNode(val)
            if not prevHead:
                prevHead = current
            if prev:
                prev.next = current
            prev = current
        return prevHead
